lib/Glimmer.pm

```diff
@@ -0,0 +1,494 @@
+	if ($motifs[$i] =~ m/^[acgt]:((\s+\d+){$motif_width})/) {
+	    $last = $i;
+	}
+    }
+    $#motifs = $last;
+    
+    my $tmp_matrix = "$tmp_prefix.pwm";
+    open( MATRIX, ">$tmp_matrix") || die "Could not write-open $tmp_matrix";
+    print MATRIX  "$motif_width\n";
+    foreach my $line (@motifs) {
+	chomp $line;
+	my @fields = split /\s+/, $line;
+	my $base   = substr((shift @fields), 0, 1);
+	print MATRIX ($base, (map { sprintf(qq(%7u), $_) } @fields), qq(\n));
+    }
+    close(MATRIX)    || die "Could not close $tmp_matrix";
+    
+    (-s $tmp_matrix) || die "Could not construct PWM --- empty file $tmp_matrix";
+    
+    
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Re-call PEGs using ICM and PWM
+    #-----------------------------------------------------------------------
+    print STDERR "Re-calling PEGs using trained ICM\n" if $verbose;
+    
+    run($verbose, "glimmer3 $glimmeropts $genetic_code_switch -b $tmp_matrix -P $start_freqs $contigs_file $tmp_model $tmp_prefix");
+    
+    #+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    #... Parse and write final predictions
+    #    NOTE: Does not yet handle '-skip' option
+    #-----------------------------------------------------------------------
+    my $tmp_predictions = "$tmp_prefix.predict";
+    print STDERR "\nParsing $tmp_predictions\n"
+	if $verbose;
+    
+    open(PREDICT,  "<$tmp_predictions")
+	|| die "Could not read-open $tmp_predictions";
+    
+    my $fid_num = 0;
+
+    my @output;
+
+    $contig_id  = qq();
+    while(defined($entry = <PREDICT>)) {
+	chomp $entry;
+
+	if ($entry =~ m/^>(\S+)/) {
+	    $contig_id = $1;
+	    next;
+	}
+	elsif ($entry =~ m/^(\S+)\s+(\d+)\s+(\d+)/) {
+	    my ($id, $beg, $end) = ($1, $2, $3);
+	    
+	    if ($beg && $end) {
+		++$fid_num;
+		my $fid = $id_prefix . $fid_num;
+		# print STDOUT (qq($fid\t), join(qq(_), ($contig_id, $beg, $end)), qq(\n));
+		my $dna = get_dna_seq($contig_id, $beg, $end, $len_of, $seq_of);
+		push(@output, [$fid, $contig_id, $beg, $end, $dna]);
+	    }
+	    else {
+		die "Error in $tmp_predictions line $.: $entry";
+	    }
+	}
+	else {
+	    die "Could not parse prediction $.: $entry";
+	}
+    }
+    
+    if (not $verbose) {
+	run($verbose, "rm -f $tmp_prefix.*");
+    }
+    else {
+	print STDERR "\nKeeping tmp files\n\n";
+    }
+    return \@output;
+}    
+    
+    ########################################################################
+sub get_dna_seq {
+    my ($contig, $beg, $end, $len_of, $seq_of) = @_;
+    my $dna = qq();
+    
+    my $contig_len = $len_of->{$contig};
+    my $contig_seq = $seq_of->{$contig};
+    
+    return undef if (&max($beg, $end) > $contig_len);
+    return undef if (&min($beg, $end) < 1);
+    
+    if ($beg < $end) {
+	$dna = substr($contig_seq, ($beg-1), ($end+1-$beg));
+    }
+    else {
+	$dna = substr($contig_seq, ($end-1), ($beg+1-$end));
+	$dna = $ { &rev_comp(\$dna) };
```

Comment:
???

why pass a reference to `$dna` instead of `$dna` itself? You can save yourself some scalar dereferencing by just passing the variable.